Re-write InputStreamSource as GraphStage

chbatey opened the pull request:
Part of https://github.com/akka/akka/issues/26187

chbatey committed:
Wip

chbatey committed:
We-write InputStreamPublisher as a GraphStage

chbatey committed:
Merge branch 'remove-actor-publisher' of github.com:chbatey/akka into remove-actor-publisher

chbatey committed:
Handle exception in factory

chbatey committed:
Mima

chbatey committed:
revert delete of inbox doc test

chbatey committed:
Merge branch 'master' into remove-actor-publisher

chbatey committed:
Delete pesky file

helena commented on akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSourceSpec.scala:
```diff
@@ -7,51 +7,88 @@ package akka.stream.io
         })
 
-      Await.result(f.takeWithin(5.seconds).runForeach(it => ()), 10.seconds)
+      Await.result(f.takeWithin(1.seconds).runForeach(it => ()), 2.seconds)
```
👍

helena commented on akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala:
```diff
@@ -45,9 +44,9 @@ object StreamConverters {
-      new InputStreamSource(in, chunkSize, DefaultAttributes.inputStreamSource, sourceShape("InputStreamSource")))
+  def fromInputStream(in: () => InputStream, chunkSize: Int = 8192): Source[ByteString, Future[IOResult]] = {
+    Source.fromGraph(new InputStreamGraphStage(in, chunkSize))
+  }
```
Need the brackets if we didn't have them before? IMO cleaner w/out.

helena commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamGraphStage.scala:
```diff
@@ -0,0 +1,85 @@
+
+      override def onPull(): Unit = {
+        tryRead()
+      }
```
Can be a one-liner, no brackets

helena commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamGraphStage.scala:
```diff
@@ -0,0 +1,85 @@
+          mat.trySuccess(IOResult(readBytesTotal, Success(Done)))
+        } catch {
+          case ex: Exception =>
+            mat.success(IOResult(readBytesTotal, Failure(ex)))
```
does seem like an odd convention to say x.success(..failure)

helena replied:
Perhaps we should add `mat.failure` rather than `mat.success` and creating a `new Failure` just to wrap an exception and signal this completed.

johanandren replied:
Yes, it is a weird consequence of how the API is defined to be able to provide how many bytes was read before it failed. I guess it is better that it is consistent though, never failing the future matval.

helena reviewed (commented):
Just a few minor questions

johanandren commented on akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSourceSpec.scala:
```diff
@@ -7,51 +7,88 @@ package akka.stream.io
+        }
+
+      }
+    }
```
Why not just use `ByteArrayInputStream`?

johanandren commented on akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSourceSpec.scala:
```diff
@@ -7,51 +7,88 @@ package akka.stream.io
+    }
 
-      f.futureValue should ===(ByteString("abc"))
+    "return IOResult with failire if creation fails" in {
```
failure

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamGraphStage.scala:
```diff
@@ -0,0 +1,85 @@
+        try {
+          if (inputStream != null)
+            inputStream.close()
+          mat.trySuccess(IOResult(readBytesTotal, Success(Done)))
```
I'd recommend we complete it successfully or with an error elsewhere and complete with a `AbruptStageTerminationException` here like other operators with a lifecycled-resource. (Would be good to cover that with a test also, run the stream and kill the materialiser while it is running to exercise)

chbatey replied:
Yep i'll do that :+1:

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamGraphStage.scala:
```diff
@@ -0,0 +1,85 @@
+/**
+ * INTERNAL API
+ */
+private[akka] class InputStreamGraphStage(factory: () => InputStream, chunkSize: Int)
```
If you liked it, make it final. Also `@InternalApi`

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamGraphStage.scala:
```diff
@@ -0,0 +1,85 @@
+    val logic = new GraphStageLogicWithLogging(shape) with OutHandler {
+      val buffer = new Array[Byte](chunkSize)
+      var readBytesTotal = 0L
+      var inputStream: InputStream = _
```
private up these just for good measure?

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamGraphStage.scala:
```diff
@@ -0,0 +1,85 @@
+            mat.success(IOResult(readBytesTotal, Failure(ex)))
+        }
+      }
+    }
```
Missing an `onDownstreamFinish` dealing with downstream cancelling? What does that mean for the matval, a `AbruptIOTerminationException` (which confuses the matval story even further)?

chbatey replied:
I guess `AbruptIOTerminationException` reserving the IOResult failure for failures with the input stream

chbatey replied:
Hrmm but then what is the exception inside the AbsruptIOTerminationException? :(

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamGraphStage.scala:
```diff
@@ -0,0 +1,85 @@
+    }
+    (logic, mat.future)
+  }
+}
```
Provide a nicer `toString`?

chbatey replied:
:+1:

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamGraphStage.scala:
```diff
@@ -0,0 +1,85 @@
+private[akka] class InputStreamGraphStage(factory: () => InputStream, chunkSize: Int)
+    extends GraphStageWithMaterializedValue[SourceShape[ByteString], Future[IOResult]] {
+
+  val out: Outlet[ByteString] = Outlet("InputStreamSource")
```
`InputStreamSource.out`

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamGraphStage.scala:
```diff
@@ -0,0 +1,85 @@
+ * INTERNAL API
+ */
+private[akka] class InputStreamGraphStage(factory: () => InputStream, chunkSize: Int)
+    extends GraphStageWithMaterializedValue[SourceShape[ByteString], Future[IOResult]] {
```
Original stage had a `chunksize > 0` check

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamGraphStage.scala:
```diff
@@ -0,0 +1,85 @@
+            val readBytes = inputStream.read(buffer)
+            readBytes match {
+              case -1 =>
+                completeStage()
```
We'd also know if we reached the end if the size was less than the buffer, no?

johanandren replied:
You are right, I must have misremembered that.

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamGraphStage.scala:
```diff
@@ -0,0 +1,85 @@
+                readBytesTotal += readBytes
+                push(out, ByteString.fromArray(buffer, 0, readBytes))
+            }
+            tryRead()
```
out will never be available here, we either completed the stage or just pushed, so no need to `tryRead` unless we start buffering for next pull.

chbatey committed:
Deprecate IOResult.failure and introduce IOOperationIncompleteException

After some dicussion with @johanandren, @raboof and @2m about the
confusion of the materialised value of the IO streams to complete
even if there is an exception (with the exception in the IOResult)
this now changes it to:

* Always fail the materialised value with IOOperationIncompleteException
if the InputStream is not fully read
* Deprecate failure in IOResult so it is always set to Success(Done)
* Stop using AbrubtIOTerminationException as the inner IOResult also
contains an exception causing confusion

chbatey committed:
Mima

chbatey commented:
@johanandren, @raboof I think this is what we concluded:

* Always fail the materialised value with IOOperationIncompleteException
if the InputStream is not fully read
* Deprecate failure in IOResult so it is always set to Success(Done)
* Stop using AbrubtIOTerminationException as the inner IOResult also
contains an exception causing confusion

johanandren commented on akka-stream/src/main/scala/akka/stream/IOResult.scala:
```diff
@@ -56,6 +66,18 @@ object IOResult {
+ */
+final case class IOOperationIncompleteException(count: Long, cause: Throwable)
+    extends RuntimeException(cause)
+    with NoStackTrace
```
Let's keep the stack trace?

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala:
```diff
@@ -0,0 +1,107 @@
+      }
+
+      setHandler(out, this)
+    }
```
postStop with the abrupt stage termination etc. see `MaybeSource` for an example.

chbatey replied:
Doh forgot about that!

johanandren reviewed (commented):
👍 on the matval changes

chbatey committed:
Review feedback

johanandren commented on akka-stream/src/main/scala/akka/stream/IOResult.scala:
```diff
@@ -56,6 +66,16 @@ object IOResult {
+ * @param count The number of bytes read/written up until the error
+ * @param cause cause
+ */
+final case class IOOperationIncompleteException(count: Long, cause: Throwable) extends RuntimeException(cause)
```
Does it really make sense to make it a case class? (bincomp with copy etc etc)

raboof commented on akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSourceSpec.scala:
```diff
@@ -4,54 +4,110 @@
+  private def inputStreamFor(bytes: Array[Byte]): InputStream =
+    new ByteArrayInputStream(bytes)
 
-    "not signal when no demand" in {
```
Indeed that didn't seem to test anything

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala:
```diff
@@ -0,0 +1,120 @@
+      private val buffer = new Array[Byte](chunkSize)
+      private var readBytesTotal = 0L
+      private var inputStream: InputStream = _
+      private var closed = false
```
`closed` seems to be pretty much a flag that is set when completing the promise, so maybe we can make `closed` a `def` checking if the promise is completed instead?

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala:
```diff
@@ -0,0 +1,120 @@
+import scala.util.control.NonFatal
+
+case class DownstreamFinishedException()
+    extends RuntimeException("Downstream Finished before stream was fully consumed")
```
final, non case class?

chbatey replied:
Removed after @raboof 's comment

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala:
```diff
@@ -0,0 +1,120 @@
+        closeInputStream()
+        mat.success(IOResult(readBytesTotal))
+        completeStage()
+      }
```
If `inputStream.close()` throws we will get an exception trying to call `success` on an already completed promise here and then fail try to close the input stream again in `failStream` in the catch in `onPull`

chbatey replied:
:+1:

johanandren reviewed (commented):
Some thought's on the error handling

raboof commented on akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSourceSpec.scala:
```diff
@@ -4,54 +4,110 @@
+        .futureValue shouldEqual IOOperationIncompleteException(0, fail)
+    }
+
+    "fail materialized value if downstream doesn't read all of it" in {
```
I'm not sure this is the desired behavior: it can be perfectly valid for downstream not to read everything, right? 

The 'old' `postStop` also did `completionPromise.trySuccess(IOResult(readBytesTotal, Success(Done)))` regardless of whether the stream was completely consumed or not.

raboof commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala:
```diff
@@ -0,0 +1,120 @@
+      override def onDownstreamFinish(): Unit = {
+        if (!closed) {
+          closeInputStream()
+          mat.tryFailure(IOOperationIncompleteException(readBytesTotal, DownstreamFinishedException()))
```
```suggestion
          mat.trySuccess(IOResult(readBytesTotal))
```

chbatey replied:
:+1: will commit this with the test change

raboof commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala:
```diff
@@ -0,0 +1,120 @@
+      private val buffer = new Array[Byte](chunkSize)
+      private var readBytesTotal = 0L
+      private var inputStream: InputStream = _
+      private var closed = false
```
Do we need this variable? Couldn't we check `inputStream.isClosed()`? If we need this variable, should we set it to 'true' when we actively close the `inputStream`?

chbatey replied:
Changed it to rely on the mat value

raboof reviewed (commented):
Overall looks good, not sure about the behavior when downstream finishes without failure

chbatey committed:
Review feedback

chbatey committed:
Merge branch 'master' into remove-actor-publisher

chbatey commented:
@raboof @johanandren when you have time it would be good to take a look at this again and if it is okay i'll re-write the others

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala:
```diff
@@ -0,0 +1,116 @@
+        } catch {
+          case NonFatal(t) =>
+            failStream(t)
+            failStage(t)
```
failStream already includes a `failStage`

chbatey replied:
`failStream` completes the mat value but doesn't call `failStage` unless an exception is thrown closing the input stream

johanandren replied:
Ah, sorry. Correct!

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala:
```diff
@@ -0,0 +1,116 @@
+        }
+      }
+
+      override def onPull(): Unit = if (isAvailable(out)) {
```
I don't think there is a scenario we get a `onPull` but `out` isn't available

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala:
```diff
@@ -0,0 +1,116 @@
+            case _ =>
+              readBytesTotal += readBytes
+              push(out, ByteString.fromArray(buffer, 0, readBytes))
+          }
```
Small style thing, could be a matter of taste, I think I would have matched directly on the return and capture the non `-1` value in a `readBytes` name instead of `_` and using the value from the outer scope.

johanandren reviewed (commented):
Needs a mention in the migration guide that it will fail with a different exception now.

chbatey committed:
Merge branch 'master' into remove-actor-publisher

chbatey committed:
Review feedback and migration guide note

johanandren reviewed (approved):
LGTM

patriknw commented on akka-docs/src/main/paradox/project/migration-guide-2.5.x-2.6.x.md:
```diff
@@ -208,6 +208,13 @@ A full cluster restart is required to change to Artery.
+### IOSources
+
+`[StreamConverters.fromInputStream` now always fails the materialized value in case of failure. It is no longer required
+to both check the materialized value and the `Try[Done]` inside the @apidoc[IOResult]. In casw of an IO failure
```
typo casw

raboof replied:
```suggestion
to both check the materialized value and the `Try[Done]` inside the @apidoc[IOResult]. In case of an IO failure
```

patriknw commented on akka-stream/src/main/scala/akka/stream/IOResult.scala:
```diff
@@ -56,6 +66,16 @@ object IOResult {
+ * @param count The number of bytes read/written up until the error
+ * @param cause cause
+ */
+final class IOOperationIncompleteException(val count: Long, cause: Throwable) extends RuntimeException(cause)
```
should it include a message to RuntimeException for better toString and logging?

patriknw commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala:
```diff
@@ -0,0 +1,114 @@
+            inputStream.close()
+        } catch {
+          case NonFatal(ex) =>
+            mat.tryFailure(new IOOperationIncompleteException(readBytesTotal, ex))
```
if `closeInputStream` is called from `failStream` it will use the "wrong" `cause` here

chbatey replied:
Okay, we could complete the mat value first but i think it is better to not touch the input stream after we close the mat value so let's have this take in an optional reason to use and if not specified use the new exception

patriknw commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala:
```diff
@@ -0,0 +1,114 @@
+      }
+
+      private def closeStage(): Unit = {
+        closeInputStream()
```
should failure in `closeInputStream` from here complete the `mat` with failure?

chbatey replied:
It does, but whether it should I don't feel strongly either way about. We didn't successfully close the input stream but we did read all the bytes

patriknw replied:
yeah, difficult to be sure if all was good, probably best to keep it as is, i.e. fail because something could be wrong

patriknw commented on akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala:
```diff
@@ -0,0 +1,114 @@
+
+  override def shape: SourceShape[ByteString] = SourceShape(out)
+
+  override protected def initialAttributes: Attributes = DefaultAttributes.inputStreamSource
```
ok, that has `IODispatcher`

patriknw reviewed (commented):
looking good, some minor nits

chbatey committed:
Merge branch 'master' into remove-actor-publisher

chbatey committed:
Merge branch 'remove-actor-publisher' of github.com:chbatey/akka into remove-actor-publisher

raboof reviewed (approved):
Looks good but has conflicts

chbatey committed:
Remove warnings now akka stream tests are warning free

chbatey committed:
Merge branch 'master' into remove-actor-publisher

chbatey committed:
Merge branch 'master' into remove-actor-publisher

chbatey committed:
Formatting

chbatey committed:
Fixup migration guide
